rockcraft/oci.py

```diff
@@ -320,15 +339,62 @@ def _add_layer_into_image(image_path: Path, archived_content: Path, **kwargs) ->
     _process_run(cmd + ["--history.created_by", " ".join(cmd)])
 
 
-def _archive_layer(layer_path: Path, temp_tar_file: Path) -> None:
+def _archive_layer(
+    layer_path: Path, temp_tar_file: Path, lower_rootfs: Optional[Path] = None
+) -> None:
     """Prepare new OCI layer by archiving its content into tar file.
 
     :param layer_path: path to the content to be archived into a layer
-    :param temp_tar_file: path to the temporary tar fail holding the archived content
+    :param temp_tar_file: path to the temporary tar file holding the archived content
+    :param lower_rootfs: optional path to the filesystem containing the extracted
+        base below this new layer. Used to preserve lower-level directory symlinks,
+        like the ones from Debian/Ubuntu's usrmerge.
     """
     with tarfile.open(temp_tar_file, mode="w") as tar_file:
-        # With `arcname="."` the contents are added relative to `layer_path`.
-        tar_file.add(layer_path, arcname=".")
+        for dirpath, subdirs, filenames in os.walk(layer_path):
+            # Sort `subdirs` in-place, to ensure that `os.walk()` iterates on
+            # them in sorted order.
+            subdirs.sort()
+
+            upper_subpath = Path(dirpath)
+            lower_symlink_target: Optional[Path] = None
+
+            # Handle adding an entry for the directory. We only do that if:
+            # - The directory is not the root (to skip a spurious "." entry);
+            # - The directory's name does not exist on ``lower_rootfs`` as a symlink
+            #   to another directory (like in usrmerge).
+            if upper_subpath != layer_path:
+                add_dir = True
+                relative_path = upper_subpath.relative_to(layer_path)
+
+                if lower_rootfs is not None:
+                    lower_path = lower_rootfs / relative_path
+                    emit.trace(
+                        f"Skipping {upper_subpath} because it exists as a symlink on the lower layer"
+                    )
+                    add_dir = not lower_path.is_symlink()
+                    if lower_path.is_symlink():
```

Comment:
Then we could set `is_symlink` to `lower_path.is_symlink()`, and test the newly set variable.